optuna/samplers/_tpe/_truncnorm.py

```diff
@@ -219,35 +219,44 @@ def _ndtri_exp(y: np.ndarray) -> np.ndarray:
 
 
 def ppf(q: np.ndarray, a: np.ndarray | float, b: np.ndarray | float) -> np.ndarray:
+    """
+    Compute the percent point function (inverse of cdf) at q of the given truncated Gaussian.
+
+    Namely, this function returns the value `c` such that:
+        q = \\int_{a}^{c} f(x) dx
+
+    where `f(x)` is the probability density function of the truncated normal distribution with
+    the lower limit `a` and the upper limit `b`.
+
+    More precisely, this function returns `c` such that:
+        ndtr(c) = ndtr(a) + q * (ndtr(b) - ndtr(a))
+    for the case where `a < 0`, i.e., `case_left`. For `case_right`, we flip the sign for the
+    better numerical stability.
+    """
     q, a, b = np.atleast_1d(q, a, b)
     q, a, b = np.broadcast_arrays(q, a, b)
 
     case_left = a < 0
     case_right = ~case_left
+    log_mass = _log_gauss_mass(a, b)
 
-    def ppf_left(q: np.ndarray, a: np.ndarray, b: np.ndarray) -> np.ndarray:
-        log_Phi_x = _log_sum(_log_ndtr(a), np.log(q) + _log_gauss_mass(a, b))
+    def ppf_left(q: np.ndarray, a: np.ndarray, b: np.ndarray, log_mass: np.ndarray) -> np.ndarray:
+        log_Phi_x = _log_sum(_log_ndtr(a), np.log(q) + log_mass)
         return _ndtri_exp(log_Phi_x)
 
-    def ppf_right(q: np.ndarray, a: np.ndarray, b: np.ndarray) -> np.ndarray:
-        log_Phi_x = _log_sum(_log_ndtr(-b), np.log1p(-q) + _log_gauss_mass(a, b))
+    def ppf_right(q: np.ndarray, a: np.ndarray, b: np.ndarray, log_mass: np.ndarray) -> np.ndarray:
+        # NOTE(nabenabe): Since the numerical stability of log_ndtr is better in the left tail, we
+        # flip the side for a >= 0.
+        log_Phi_x = _log_sum(_log_ndtr(-b), np.log1p(-q) + log_mass)
         return -_ndtri_exp(log_Phi_x)
 
     out = np.empty_like(q)
+    if (q_left := q[case_left]).size:
+        out[case_left] = ppf_left(q_left, a[case_left], b[case_left], log_mass[case_left])
+    if (q_right := q[case_right]).size:
+        out[case_right] = ppf_right(q_right, a[case_right], b[case_right], log_mass[case_right])
 
-    q_left = q[case_left]
-    q_right = q[case_right]
-
-    if q_left.size:
-        out[case_left] = ppf_left(q_left, a[case_left], b[case_left])
-    if q_right.size:
-        out[case_right] = ppf_right(q_right, a[case_right], b[case_right])
-
-    out[q == 0] = a[q == 0]
-    out[q == 1] = b[q == 1]
-    out[a == b] = math.nan
-
-    return out
+    return np.select([a == b, q == 1, q == 0], [math.nan, b, a], default=out)
```

Comment:
Could `default=out` result in random, uninitialized values being returned?
It may be safer to use `default=None` and assert that the result does not contain `None`.